### PR TITLE
Added list of newly created NFTs

### PR DESCRIPTION
**Pull request by mstrug-rdx**
### Summary

Added list of newly created NFTs during the transaction to the Execution Summary.

### CODE REVIEW

File: crates/radix-engine-toolkit/src/transaction_types/interface.rs [hunk 1]
```diff
@@ -300,6 +302,29 @@ pub fn execution_summary(
         royalty_cost: receipt.fee_summary.total_royalty_cost_in_xrd,
     };
 
+    for (node_id, states_to_update) in
```
Comment (0xOmarA) [inline crates/radix-engine-toolkit/src/transaction_types/interface.rs:305]: I think that it would be better if we avoid doing any `expect_` calls in here, mainly for two reasons:

1. We would like to minimize the number of code branches where panics can happen.
2. The `TransactionTypesReceipt` type already stores the `CommitResult`. So, there is no need to perform this `expect` again.

Additionally, can the functionality to get the newly created non-fungibles be implemented as a method on the `TransactionTypesReceipt` type itself?

File: crates/radix-engine-toolkit/src/transaction_types/interface.rs [hunk 2]
```diff
@@ -300,6 +302,29 @@ pub fn execution_summary(
...
+        receipt.expect_commit_ignore_outcome().state_updates()
+    {
+        if node_id.is_global_non_fungible_resource_manager() {
```
Comment (0xOmarA) [inline crates/radix-engine-toolkit/src/transaction_types/interface.rs:308]: I think that a disadvantage of this approach is that we can't distinguish between newly minted NFTs and updates to the non-fungible data. It might be better to get this information from events due to that. Also, this might be a good test case to add, non-fungible data updates are not picked up as newly minted non-fungibles.

**Review by 0xOmarA — COMMENTED** [review]
Nice! I've left a few comments. I think that we're also missing the implementation of this in the `radix-engine-toolkit-uniffi` crate. Is this something you'd like to address in this PR, or a future one?

File: crates/radix-engine-toolkit/src/transaction_types/interface.rs (hunk 2 above)
Comment (mstrug-rdx) [reply crates/radix-engine-toolkit/src/transaction_types/interface.rs]: I've implemented logic based on events.

Comment (mstrug-rdx) [reply crates/radix-engine-toolkit/src/transaction_types/interface.rs]: Also I've implemented `radix-engine-toolkit-uniffi` part.

File: crates/radix-engine-toolkit/src/transaction_types/types.rs [hunk 3]
```diff
@@ -373,6 +375,50 @@ impl<'r> TransactionTypesReceipt<'r> {
...
+
+    pub fn new_nonfungibles(&self) -> Vec<NonFungibleGlobalId> {
+        let mut minted_id_list = Vec::new();
+        let mut burnt_id_list = Vec::new();
+        for (event_type, event_payload) in
+            self.commit_result.application_events.iter()
+        {
+            match event_type.0 {
+                Emitter::Method(node_id, ..) => {
+                    if node_id.entity_type().unwrap()
```
Comment (0xOmarA) [inline crates/radix-engine-toolkit/src/transaction_types/types.rs:387]: Instead of the `node_id.entity_type().unwrap()` it might be better to use: `if let ResourceAddress::try_from(node_id.bytes())`

```rust
match ResourceAddress::try_from(node_id.as_bytes()) {
    Some(address) if !address.is_fungible() => {
        /* Some logic */
    }
    _ => {}
}
```

Generally, I would like us to avoid unwraps in all of the code paths and the above described flow allows us to avoid them.

File: crates/radix-engine-toolkit/src/transaction_types/types.rs [hunk 4]
```diff
@@ -373,6 +375,50 @@ impl<'r> TransactionTypesReceipt<'r> {
             )
             .collect()
     }
+
+    pub fn new_nonfungibles(&self) -> Vec<NonFungibleGlobalId> {
```
Comment (0xOmarA) [inline crates/radix-engine-toolkit/src/transaction_types/types.rs:379]: Minor: can we rename to `new_non_fungibles`?

File: crates/radix-engine-toolkit/src/transaction_types/types.rs [hunk 5]
```diff
@@ -373,6 +375,50 @@ impl<'r> TransactionTypesReceipt<'r> {
...
+                            }
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+
+        minted_id_list.retain(|item| !burnt_id_list.contains(item));
+        minted_id_list
```
Comment (0xOmarA) [inline crates/radix-engine-toolkit/src/transaction_types/types.rs:420]: Can we return a `HashSet<NonFungibleGlobalId>` here and keep using `Vec` in uniffi?

File: crates/radix-engine-toolkit-uniffi/src/transaction/manifest.rs [hunk 6]
```diff
@@ -792,6 +793,20 @@ impl ExecutionSummary {
...
+                .into_iter()
+                .map(|item| {
+                    NonFungibleGlobalId::from_parts(
+                        Arc::new(Address::unsafe_from_raw(
+                            item.resource_address().into_node_id(),
+                            network_id,
+                        )),
+                        NonFungibleLocalId::from(item.local_id().clone()),
+                    )
+                    .unwrap()
```
Comment (0xOmarA) [inline crates/radix-engine-toolkit-uniffi/src/transaction/manifest.rs:807]: Let's avoid this unwrap by doing the following:

```rust
newly_created_non_fungibles: native
  .newly_created_non_fungibles
  .into_iter()
  .map(|item| {
      NonFungibleGlobalId::from_parts(
          Arc::new(Address::unsafe_from_raw(
              item.resource_address().into_node_id(),
              network_id,
          )),
          NonFungibleLocalId::from(item.local_id().clone()),
      )
  })
  .collect::<Result<_, _>>()?,
```

**Review by 0xOmarA — COMMENTED** [review]
Looks good, thanks for making the changes! There are a few small changes required.

Can we also add a test case that tests that a mint followed by a burn results in an empty set/vec of newly minted NFTs?

File: crates/radix-engine-toolkit-uniffi/src/transaction/manifest.rs (hunk 6 above)
Comment (mstrug-rdx) [reply crates/radix-engine-toolkit-uniffi/src/transaction/manifest.rs:807]: This unwrap should never panic because we are converting valid `NativeNonFungibleGlobalId`, in normal operation this always should succeed. In case it fails it means `CoreExecutionSummary` is malformed so I think we should change `ExecutionSummary::from_native()` return type to `Result` and return error in that case. What is your opinion?

Comment (0xOmarA) [reply crates/radix-engine-toolkit-uniffi/src/transaction/manifest.rs:807]: Agreed that this is a path that should realistically never panic. However, since this function already returns a `Result` then it makes sense to propagate all `Result`s.

**Review by 0xOmarA — APPROVED** [review]
